### PR TITLE
Fixed bare `raise` and exception chaining when a handler raises an exception

### LINKED ISSUE #70: bare raise in catch behaves differently to bare raise in native except*
with the backport:

```python
import sys
import exceptiongroup


def raise_exc():
    raise ExceptionGroup("bad", [ValueError(), ValueError(), TypeError()])

def handle(eg):
    raise

def noop(*args, **kwargs):
    pass

with exceptiongroup.catch({ValueError: handle, TypeError: noop}):
    raise_exc()
```

I get:

```pytb
  + Exception Group Traceback (most recent call last):
  |   File "/home/graingert/.virtualenvs/testing311/lib/python3.11/site-packages/exceptiongroup/_catch.py", line 52, in handle_exception
  |     handler(matched)
  |   File "/home/graingert/projects/dask/backport.py", line 15, in <module>
  |     raise_exc()
  |   File "/home/graingert/projects/dask/backport.py", line 6, in raise_exc
  |     raise ExceptionGroup("bad", [ValueError(), ValueError(), TypeError()])
  | ExceptionGroup: bad (3 sub-exceptions)
  +-+---------------- 1 ----------------
    | ValueError
    +---------------- 2 ----------------
    | ValueError
    +---------------- 3 ----------------
    | TypeError
    +------------------------------------
```

with the native code:

```python
import sys
import exceptiongroup


def raise_exc():
    raise ExceptionGroup("bad", [ValueError(), ValueError(), TypeError()])

try:
    raise_exc()
except* ValueError as eg:
    raise
except* TypeError:
    pass
```

I get:

```pytb
  + Exception Group Traceback (most recent call last):
  |   File "/home/graingert/projects/dask/native.py", line 9, in <module>
  |     raise_exc()
  |   File "/home/graingert/projects/dask/native.py", line 6, in raise_exc
  |     raise ExceptionGroup("bad", [ValueError(), ValueError(), TypeError()])
  | ExceptionGroup: bad (2 sub-exceptions)
  +-+---------------- 1 ----------------
    | ValueError
    +---------------- 2 ----------------
    | ValueError
    +------------------------------------
```

### PR DESCRIPTION
**Pull request by agronholm**
Fixes #70.

### CODE REVIEW

File: tests/test_catch.py [hunk 1]
```diff
@@ -153,6 +153,21 @@ def handler(exc):
             raise ExceptionGroup("booboo", [ValueError("bar")])
 
 
+def test_catch_handler_reraises():
+    def handler(exc):
+        raise
```
Comment (graingert) [inline tests/test_catch.py:158]: Can you add a test with `raise Exception` and make sure the `__context__` is set to the split exception group

Comment (agronholm) [reply tests/test_catch.py:160]: The `__context__` value is the same as `__cause__` on py3.11 though?

Comment (agronholm) [reply tests/test_catch.py:160]: So it seems that in the backport, `__context__` is set to the split exception group which doesn't match the py3.11 test which I just augmented.

Comment (agronholm) [reply tests/test_catch.py:160]: I made sure that the back-port works the same way as py3.11 does.

Comment (agronholm) [reply tests/test_catch.py:160]: As usual, I misread your comment. I'll make sure to add such a test.

Comment (agronholm) [reply tests/test_catch.py:160]: Well, this was interesting. Contrary to my expectations, the exception group in `__context__` on py3.11 was a _copy_ of the original, so `assert exc.value.__context__ is excgrp` failed.

Comment (agronholm) [reply tests/test_catch.py:168]: I'm not sure that it's essential to copy the exception group. I stopped just short of doing that, as I'm not sure it's worth the effort to change the context to a copy.

**Pull request merged**